File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-1.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        # mid > right：说明 mid 在“大区间”，断点在右边 → 去右边找：left = mid + 1
        # mid <= right： 说明右边是有序小区间，断点在左边（包括 mid）：right = mid
        
        left = 0

        right = len(nums) - 1

        while left < right:
            mid = (left + right) // 2

            if (nums[mid] > nums[right]):
                left = mid + 1
            else: 
                right = mid
        return nums[left]

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-1.js
class Solution {
    /**
     * @param {string} s
     * @param {number} k
     * @return {number}
     */
    characterReplacement(s, k) {
        const count = {};

        // 窗口问题， 需要修改次数 =windowSize - maxFreq
        let left = 0
        let maxFreq = 0;
        let maxLength = 0;

        for (let right = 0; right < s.length; right++) {
            const char = s[right];
            // 字符计数
            count[char] = (count[char] || 0) + 1;
            // 最大相同字符数
            maxFreq = Math.max(maxFreq, count[char])

            // 如果窗口大小 - maxFreq > k，说明k不能覆盖，则left指针右移
            while (
                (right - left + 1) - maxFreq > k
            ) {
                count[s[left]]--;
                left++;
            }

            maxLength = Math.max(
                maxLength,
                right - left + 1
            );
        }

        return maxLength

    }
}
